Source/WebCore/rendering/updating/RenderTreeUpdater.cpp

```diff
@@ -314,6 +314,9 @@ void RenderTreeUpdater::updateElementRenderer(Element& element, const Style::Ele
     auto elementUpdateStyle = RenderStyle::cloneIncludingPseudoElements(*elementUpdate.style);
 
     bool shouldTearDownRenderers = elementUpdate.change == Style::Change::Renderer && (element.renderer() || element.hasDisplayContents() || element.displayContentsChanged());
+    bool isTopLayerHiddenByContentVisibility = element.isInTopLayer() && elementUpdate.style->effectiveSkipsContent() && renderTreePosition().parent().style().effectiveSkipsContent();
+    if (!shouldTearDownRenderers && isTopLayerHiddenByContentVisibility)
+        shouldTearDownRenderers = true;
```

Comment:
This is probably not the right place to do this. Would testing in `Element::rendererIsNeeded` give the right behavior?

Reply:
I tried, but unfortunately I think rendererIsNeeded would get too complicated or even couldn't do it (need to take parent/parent style into account). The problem is that we have two situations:
1. The dialog has content-visibility: hidden set on it, this means it should render but not its children.
2. The dialog is in a content-visibility: hidden subtree and in this case there is no point in even creating a renderer for it.
I did notice that my previous patch did too much work for the dialog children when content-visibility: hidden applies, so I adjusted the mayHaveRenderedDescendants logic. I think overall RenderTreeUpdater has the best info to make the right decision about content-visibility: hidden and dialogs.